### PR TITLE
Add support for JSON schema

### PR DESCRIPTION
**Pull request by yaodingyd**
Fix https://github.com/sindresorhus/electron-store/issues/16

Use epoberezkin/ajv for JSON schema validation

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -139,6 +156,8 @@ class Conf {
 				} catch (_) {}
 			}
 
+			this.validate(JSON.parse(data));
```
Comment (sindresorhus) [inline index.js]: No point in doing `JSON.parse` twice.

File: index.js [hunk 2]
```diff
@@ -41,6 +42,11 @@ class Conf {
 			options.cwd = envPaths(options.projectName, {suffix: options.projectSuffix}).config;
 		}
 
+		if (options.schema && options.fileExtension === 'json') {
```
Comment (sindresorhus) [inline index.js]: It should work with any extension.

File: test.js [hunk 3]
```diff
@@ -355,3 +355,36 @@ test('doesn\'t write to disk upon instanciation if and only if the store didn\'t
 	exists = fs.existsSync(conf.path);
 	t.is(exists, true);
 });
+
+test('schema - invalid set', t => {
+	const schema = {
+		$schema: 'http://json-schema.org/draft-07/schema#',
+		type: 'object',
+		properties: {
```
Comment (sindresorhus) [inline test.js]: We should set these top-level meta things for the user. The user should only specify their schema as:

```js
{
	foo: {
		type: 'string'
	}
}
```

File: readme.md [hunk 4]
```diff
@@ -120,6 +120,13 @@ You can pass an empty string to remove the suffix.
 
 For example, on macOS, the config file will be stored in the `~/Library/Preferences/foo-nodejs` directory, where `foo` is the `projectName`.
 
+#### schema
+
+type: `Object`<br>
+Default: `undefined`
+
+JSON Schema to validate your config.
```
Comment (sindresorhus) [inline readme.md]: "JSON schema" should link to docs about it.

**Comment by sindresorhus** [comment]
You need to consider how the `defaults` option will work with schema defaults. There's overlap. https://github.com/epoberezkin/ajv#assigning-defaults

**Comment by sindresorhus** [comment]
This PR is a bit thin considering the expected work for the bounty.

**Comment by sindresorhus** [comment]
The readme documentation should have a full realistic example of how it should be used.

**Comment by sindresorhus** [comment]
Also go through all the options in https://github.com/epoberezkin/ajv#validation-and-reporting-options and carefully consider which one to use or not.

File: index.js [hunk 5]
```diff
@@ -56,6 +62,17 @@ class Conf {
 		}
 	}
 
+	validate(data) {
+		if (this.validator) {
```
Comment (sindresorhus) [inline index.js]: Would be more readable with an early-return here.

**Comment by sindresorhus** [comment]
We should at least have these options: https://github.com/wulkano/kap/blob/01a557f6616108423f9e46718d48596a22698480/main/plugin.js#L48-L53

File: index.js [hunk 6]
```diff
@@ -41,6 +42,11 @@ class Conf {
 			options.cwd = envPaths(options.projectName, {suffix: options.projectSuffix}).config;
 		}
 
+		if (options.schema && options.fileExtension === 'json') {
+			const ajv = new Ajv();
+			this.validator = ajv.compile(options.schema);
```
Comment (sindresorhus) [inline index.js]:
```suggestion
			this._validator = ajv.compile(options.schema);
```

File: index.js [hunk 7]
```diff
@@ -56,6 +62,17 @@ class Conf {
 		}
 	}
 
+	validate(data) {
```
Comment (sindresorhus) [inline index.js]:
```suggestion
	_validate(data) {
```

File: test.js [hunk 8]
```diff
@@ -355,3 +355,36 @@ test('doesn\'t write to disk upon instanciation if and only if the store didn\'t
...
+			foo: {
+				type: 'string'
+			}
+		}
+	};
+	const conf = new Conf({cwd: tempy.directory(), schema});
+	conf.set('foo', fixture);
+	t.is(conf.get('foo'), fixture);
+	t.throws(() => conf.set('foo', 1), 'Config is not valid according to schema:.foo should be string;');
```
Comment (sindresorhus) [inline test.js]: The error message is missing a space.

**Comment by sindresorhus** [comment]
Can you fix the merge conflict?

**Comment by yaodingyd** [comment]
I go through all the options for ajv:
1. We would only use basic and standard type and formats, no `$data` references, no custom `formats`;
2. We don't use referenced schema options(`schemaId`, `missingRefs`, `extendRefs` or `loadSchema`);
3. Except to `useDefault`, we don't want to modify validated data(`removeAdditional` or `coerceTypes`);
4. We don't use any advanced options except `errorDataPath`, which is good for reporting errors.
So I think to use the same options makes sense.

File: readme.md [hunk 9]
```diff
@@ -225,6 +232,33 @@ const config = new Conf({
 });
 ```
 
+### How can I use `schema` option to validate my config?
+
+Under the hood, JSON Schema validator [ajv](https://github.com/epoberezkin/ajv) is used to validate your config. We use
```
Comment (sindresorhus) [inline readme.md]: Don't hard-wrap

File: readme.md [hunk 10]
```diff
@@ -225,6 +232,33 @@ const config = new Conf({
...
+[JSON Schema draft-07](http://json-schema.org/latest/json-schema-validation.html) and support all [validation keywords](https://github.com/epoberezkin/ajv/blob/master/KEYWORDS.md) and [formats](https://github.com/epoberezkin/ajv#formats).
+
+You should define your schema as an object where each key is the name of your data's property and each value is a JSON schema used to validate that property. See more [here](https://json-schema.org/understanding-json-schema/reference/object.html#properties).
+
+Example:
+
+```js
+const schema = {
+	'foo': {
```
Comment (sindresorhus) [inline readme.md]: No need to quote the keys.

File: readme.md [hunk 11]
```diff
@@ -225,6 +232,33 @@ const config = new Conf({
...
+		type: 'number',
+		maximum: '100',
+		minimum: '1',
```
Comment (sindresorhus) [inline readme.md]: Shouldn't these be numbers?

File: readme.md [hunk 12]
```diff
@@ -225,6 +232,33 @@ const config = new Conf({
...
+	'bar': {
+		type: 'string',
+		format: 'url'
+	}
+};
+
+const config = new Conf({schema});
+```
+
+**Please be aware that your schema `default` will overwrite the corresponding property specified in `defaults` option in Conf.**
```
Comment (sindresorhus) [inline readme.md]: It should be mentioned in the `defaults` options that it's moot if you set the `schema` option.

File: test.js [hunk 13]
```diff
@@ -434,3 +434,33 @@ test('`clearInvalidConfig` option - valid data', t => {
...
+test('schema - invalid set', t => {
+	const schema = {
+		foo: {
+			type: 'string'
+		}
+	};
+	const conf = new Conf({cwd: tempy.directory(), schema});
+	conf.set('foo', fixture);
+	t.is(conf.get('foo'), fixture);
+	t.throws(() => conf.set('foo', 1), 'Config is not valid according to schema: .foo should be string;');
```
Comment (sindresorhus) [inline test.js]: The error should be:

```
Config schema violation: `foo` should be string
```

Not that there should be not `.` in front of `foo`. And it should be wrapped in backticks, and no `;` at the end.

File: test.js [hunk 14]
```diff
@@ -434,3 +434,33 @@ test('`clearInvalidConfig` option - valid data', t => {
 	conf.set('foo', 'bar');
 	t.deepEqual(conf.store, {foo: 'bar'});
 });
+
+test('schema - should be an object', t => {
+	const schema = 'object';
+	t.throws(() => new Conf({cwd: tempy.directory(), schema}), 'Schema option must be an object.');
```
Comment (sindresorhus) [inline test.js]: Don't use inline arrow function for `t.throws`.

File: test.js [hunk 15]
```diff
@@ -434,3 +434,33 @@ test('`clearInvalidConfig` option - valid data', t => {
...
+		foo: {
+			type: 'string'
+		}
+	};
+	const conf = new Conf({cwd: tempy.directory(), schema});
+	conf.set('foo', fixture);
+	t.is(conf.get('foo'), fixture);
+	t.throws(() => conf.set('foo', 1), 'Config is not valid according to schema: .foo should be string;');
+});
+
```
Comment (sindresorhus) [inline test.js]: Can you add a test for the error when there are multiple violations, to ensure it shows both errors and that it shows it correctly formatted.

File: readme.md (hunk 12 above)
Comment (sindresorhus) [reply readme.md]: Something like this:

> If you use the `schema` option, you can just set the `default` key there for values instead of using this option.

File: index.js [hunk 16]
```diff
@@ -79,6 +99,19 @@ module.exports = class Conf {
 		}
 	}
 
+	_validate(data) {
+		if (!this._validator) {
+			return true;
```
Comment (sindresorhus) [inline index.js]: Why are you returning `true` here? You're not using it anywhere.

File: readme.md [hunk 17]
```diff
@@ -225,6 +232,33 @@ const config = new Conf({
 });
 ```
 
+### How can I use `schema` option to validate my config?
```
Comment (sindresorhus) [inline readme.md]: This should just be documented where the `schema` option is defined.

File: test.js [hunk 18]
```diff
@@ -434,3 +434,33 @@ test('`clearInvalidConfig` option - valid data', t => {
...
+		foo: {
+			type: 'string'
+		}
+	};
+	const cwd = tempy.directory();
+
+	const conf = new Conf({cwd, schema});
+	fs.writeFileSync(path.join(cwd, 'config.json'), JSON.stringify({foo: 1}));
+	t.throws(() => conf.get('foo'), 'Config is not valid according to schema: .foo should be string;');
+});
```
Comment (sindresorhus) [inline test.js]: The `schema` option needs more tests:

- Valid config with the schema option.
- Using `default` in the schema.
- `default` in schema overwriting property in `defaults` option.
- Advanced schema with many different types and combinations of attributes.
- What else?...

**Comment by yaodingyd** [comment]
I was wrong about the overwrite order before: actually we apply Conf's defaults first, thus schema's default would not be used because value has already be set. I have updated doc accordingly.

File: readme.md [hunk 19]
```diff
@@ -145,6 +147,38 @@ You can pass an empty string to remove the suffix.
...
+		minimum: 1,
+		default: 50
+	},
+	bar: {
+		type: 'string',
+		format: 'url'
+	}
+};
+
+const config = new Conf({schema});
```
Comment (sindresorhus) [inline readme.md]: Can you include an example of setting an invalid value and the error message, so people get a feel of how it works?

File: readme.md [hunk 20]
```diff
@@ -225,7 +259,6 @@ const config = new Conf({
 });
 ```
 
-
```
Comment (sindresorhus) [inline readme.md]: Don't do unrelated changes.

**Comment by sindresorhus** [comment]
Can you add the option to https://github.com/sindresorhus/conf/blob/master/index.d.ts and https://github.com/sindresorhus/conf/blob/master/index.test-d.ts ?

And fix the merge conflict.

**Comment by sindresorhus** [comment]
Looks good now :)
